### PR TITLE
Allow collection paths with wildcard parameters (e.g. /users/{id}/tasks}

### PR DESCRIPTION
**Pull request by samusenkoiv**
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #884
| License       | MIT
| Doc PR        | -

This is what I got so far. It is not covered by tests yet.
I just wanted to confirm if you consider this place as an appropriate one for these changes. 
@dunglas what do you think?

### CODE REVIEW

**Comment by dunglas** [comment]
The IRI generator must work even in a non HTTP context (in CLI for instance), RequestStack cannot be a hard dependency.
I don't understand how blindly passing current parameters can help (for instance when generating IRI for resources not related with the currently related collection). Can you elaborate a bit more?

**Comment by dunglas** [comment]
What we can probably do is finding a convention to populate router parameters using data of the related resource.

Example: /foo/{foo_id}/{id}

```['foo_id' => $currentEntity->getFoo()->getId()]```.

However, I'm not sure it's worth it (it will introduce a lot of complexity for no real advantage compared to put every resources at the root).

**Comment by dunglas** [comment]
By the way, you can already do this kind of logic by decorating the `api_platform.router` service, maybe should we only add a documentation explaining how to do that.

**Comment by samusenkoiv** [comment]
That's what I thought at the first place, that RequestStack can not be used in IriCoverter as a hard dependency.
I also do not see any benefits of having routes like "/foos/{foo_id}/bars/{id}" , you'd better just go to "/bars/{id}", right?
But I do see a good thing in an ability to have routes like "/foos/{foo_id}/bars".

Moreover, there is one more problem. When I define several "get" collection operations, e.g. "/bars" and "/foos/{foo_id}/bars" the @id of the collection is always taken from the first operation.

**Comment by samusenkoiv** [comment]
@dunglas I changed the code, what do you think about it?

**Comment by samusenkoiv** [comment]
@dunglas  ping

**Comment by dunglas** [comment]
I don't get exactly what it does. Can you add a functional test to explicit it (and to be sure this is properly tested).

**Comment by samusenkoiv** [comment]
@dunglas the point is that a recourse can participate in several collections (/items, /item-owner/{item-owner-id}/items, /brand-new-items .

Some recourses even do not need to have a common collection where you can find each item of the recourse e.g. comments it makes sense to have paths like /post/{post-id}/comments or /user/{user-id}/comments, but who does ever need just /comments?

The functional tests are passed well on my local machine, what could be the reason they don't here?

**Comment by dunglas** [comment]
The failure is related to CS errors, run PHP CS Fixer.

**Comment by samusenkoiv** [comment]
@dunglas yes, it was CS errors, but along with swagger errors. I changed behavior of DocumentNormalizer to parse paths and to put placeholders {} as required path parameters.

Pardon me, but I never used these checks before, would you tell me where I can find what scrutinizer is complaining about?

**Comment by samusenkoiv** [comment]
@dunglas any feedback?

**Comment by samusenkoiv** [comment]
@dunglas or do I het it wrong and Hydra is not supposed to do something like that?

**Comment by dunglas** [comment]
I need to find some time to review this PR, sorry for the delay @samusenkoiv.

**Comment by dunglas** [comment]
Can you elaborate how URL for such patterns are generated (from where come the `{id}` placeholder)?

**Comment by samusenkoiv** [comment]
@dunglas the URL is taken from the context. While the context is created in here https://github.com/api-platform/core/blob/master/src/EventListener/SerializeListener.php#L61

File: src/Swagger/Serializer/DocumentationNormalizer.php [hunk 1]
```diff
@@ -191,6 +191,16 @@ private function updateGetOperation(\ArrayObject $pathOperation, array $mimeType
         $responseDefinitionKey = $this->getDefinition($definitions, $collection, false, $resourceMetadata, $resourceClass, $operationName);
         $pathOperation['produces'] ?? $pathOperation['produces'] = $mimeTypes;
 
+        if (!array_key_exists('parameters', $pathOperation)) {
+            $pathParameters = $this->getRequiredPathParameters($resourceMetadata, $operationName, $collection);
+            $filtersParameters = $collection ? $this->getFiltersParameters($resourceClass, $operationName, $resourceMetadata) : [];
+            $parameters = array_merge($pathParameters, $filtersParameters);
+
+            if (!empty($parameters)) {
```
Comment (dunglas) [inline src/Swagger/Serializer/DocumentationNormalizer.php]: Please use `!$parameters` (the call to empty is useless).

File: tests/Fixtures/TestBundle/Entity/RelatedDummy.php [hunk 2]
```diff
@@ -137,7 +144,7 @@ public function getRelatedToDummyFriend()
     /**
      * Set relatedToDummyFriend.
      *
-     * @param relatedToDummyFriend the value to set
+     * @param relatedToDummyFriend $relatedToDummyFriend
```
Comment (dunglas) [inline tests/Fixtures/TestBundle/Entity/RelatedDummy.php]: It should be `@param RelatedToDummyFriend  $relatedToDummyFriend`

**Comment by samusenkoiv** [comment]
@dunglas check it out

**Comment by teohhanhui** [comment]
Nested collections! This is what I've been thinking of for a long time.

But I don't think it's so simple to implement. This PR does not seem to change the IRI generation, which I think is crucial, e.g. `/posts/{post_id}/comments/{id}` instead of `/comments/{id}` to fully express the dependency.

**Comment by teohhanhui** [comment]
Also, I think this should be handled at the normalizer level, i.e. allowing to serialize / deserialize into a nested collection (as opposed to embedding directly, which stops making sense anyway when there are too many items in the collection).

Using custom routes like what's done in this PR is a band-aid solution that does not address the real issue. :smile:

**Comment by samusenkoiv** [comment]
@teohhanhui I think you got it wrong. The PR does not pretend to implement or help with nested collection.

Instead it does the following:
- Shows the _real_ @id of a collection instead of hardcoded one
- Allows to use placeholders in path, which is not something weird or extraordinary or is it?

**Comment by samusenkoiv** [comment]
@dunglas @teohhanhui so you guys think it is not worth merging the PR?

**Comment by dunglas** [comment]
I'm in favor of merging this one. What do you think @teohhanhui?

**Comment by samusenkoiv** [comment]
Oops, that was by accident

**Comment by soyuka** [comment]
so this doesn't work: `/posts/{post_id}/comments/{id}`?

I think this isn't really optimized, would you be able to show me the SQL query that is executed? Does it serialize the parent too before rendering?

Also the regex should be tested thoroughly with unit tests. Also, isn't there something better to use from the Symfony path tools or the router?

**Comment by esistgut** [comment]
This PR seems to be related to the question I asked here https://github.com/api-platform/api-platform/issues/199
Can it cover my use case?

**Comment by soyuka** [comment]
@esistgut your case is not related.

@samusenkoiv I thought a bit more about this feature. As said in my previous comment I'd love to have the corresponding query and serialization informations. 

> Allows to use placeholders in paths, which is not something weird or extraordinary or is it?

IMO this can be misleading, because the `id` in a resource is already given by ItemOperation and we should not complexify the Operation declaration. 

What we can and should do instead is to "allow operations to **retrieve** and **serialize** only the association of a parent resource". This will give the same result as your patch does but it'll go deeper and notify the data providers and the serialization about what should be fetched. I don't think it's easy to implement but I'd start with the IRI declaration and try to make the Router aware of "I want a `CollectionOperation` siblings to an `ItemOperation`". This also means that instead of trying to parse the routing parameters manually, it'll define a proper Operation that our [ApiLoader](https://github.com/api-platform/core/blob/master/src/Bridge/Symfony/Routing/ApiLoader.php) will directly load and that parameters should be interpreted by the Symfony router:

```
/api/entity/{id}/association.{_format}
/api/entity/{id}/association/{association_id}.{_format}
``` 

Where `{id}` can be one of:

- `10`
- `foo=10;bar=12`

There is a strategy in place for composite identifiers in the [Doctrine data provider](https://github.com/api-platform/core/blob/master/src/Bridge/Doctrine/Orm/ItemDataProvider.php#L129). The rest should be handled in the IriConverter and related.

**Comment by teohhanhui** [comment]
Definitely agree with @soyuka re the routing.

**Comment by teohhanhui** [comment]
@samusenkoiv

> Allows to use placeholders in paths, which is not something weird or extraordinary or is it?

I do not understand under what circumstances it would be appropriate to use? (Other than the nested collection use case you've described, because I think this is not the right way to get there.)

**Comment by soyuka** [comment]
Closing in favor of #904

**Pull request closed without merging**